B(1) = +½ set 9: Appell sequences

Parcly-Taxel opened the pull request:
#### References to other Issues or PRs

Part 9 of the changes formerly included in #23926, but initiated because of [a comment](https://github.com/sympy/sympy/pull/23989#discussion_r957861664) on #23989 suggesting a move of the combinatorial polynomial generators to the `polys` module.

#### Brief description of what is fixed or changed

The polynomial sequences associated with `bernoulli()`, `genocchi()` and `euler()` are all [Appell sequences](https://en.wikipedia.org/wiki/Appell_sequence). This means they can be generated efficiently by passing different parameters to the same piece of code, something Luschny hints at when he discusses the so-called _central Bernoulli_ and _Swiss-knife polynomials_, the latter renamed here to _André polynomials_.

This PR implements the five specific Appell sequences mentioned in the previous paragraph.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* polys
    * New `appellseqs` module generating Appell sequences: Bernoulli, central Bernoulli, Genocchi, Euler and André polynomials.
<!-- END RELEASE NOTES -->

Parcly-Taxel committed:
Initial definition of Appell sequences

Parcly-Taxel committed:
Document Appell sequences

Parcly-Taxel committed:
Link Appell sequences to corresponding continuous functions

Parcly-Taxel committed:
No, this isn't Markdown

oscarbenjamin commented on sympy/polys/appellseqs.py:
```diff
@@ -0,0 +1,156 @@
+    return seq[n]
+
+@public
+def appell_poly(n, seq, v, f, K, x=None, polys=False):
```
This function seems a bit complicated for a public function. At least it would need more explanation or a doctest or something.

I checked both of the citations and neither of them gives a clear explanation as to where the parameters `f` and `v` come from here. Presumably this characterisation of Appell sequences and this algorithm is described somewhere but it doesn't seem to be in those citations.

Also the description of seq is not completely clear e.g. should it be `Expr` or domain coefficients? Passing the domain `K` suggests that we're working in the poly domains where a list of domain elements is the DUP representation of a polynomial but I think the intention here is that a list of lists of Expr should be passed for this parameter.

Parcly-Taxel replied:
Right, perhaps I tried too hard to shoehorn everything into the same framework, eh?

oscarbenjamin replied:
I thought that the structure of the code was fine but that the `appell_poly` function didn't seem useful as a public function because it seemed unlikely that anyone would understand how to use it.

The new code does seem clearer to me though.

Parcly-Taxel committed:
Custom Appell sequence functions and a doctest

Parcly-Taxel committed:
Raw docstrings for MathJax

Parcly-Taxel committed:
Complete documentation for Appell sequences

Parcly-Taxel committed:
Fix yet more MathJax typos

oscarbenjamin commented on sympy/polys/appellseqs.py:
```diff
@@ -0,0 +1,295 @@
+
+def dup_bernoulli(n, K):
+    """Low-level implementation of Bernoulli polynomials."""
+    seq = [[1], [1, K(-1,2)]]
+    for i in range(2, n+1):
+        q = dup_integrate(dup_mul_ground(seq[-1], i, K), 1, K)
+        if i&1 ^ 1:
+            q = dup_sub_ground(q, dup_eval(q, K(1,2), K) * K(1<<(i-1), (1<<i)-1), K)
+        seq.append(q)
+    return seq[n]
```
It looks like we don't need the `seq` list at all and can just use `q = [K.one, K(-1, 2)]` at the start followed by using `q` everywhere instead of `seq[-1]`.

oscarbenjamin commented on sympy/polys/appellseqs.py:
```diff
@@ -0,0 +1,295 @@
+
+    .. [1] https://en.wikipedia.org/wiki/Bernoulli_polynomials
+    """
+    if n < 0:
+        raise ValueError("Cannot generate Bernoulli polynomial of degree %s" % n)
+    poly = DMP(dup_bernoulli(int(n), QQ), QQ)
+    if x is not None:
+        poly = Poly.new(poly, x)
+    else:
+        poly = PurePoly.new(poly, Dummy('x'))
+    return poly if polys else poly.as_expr()
```
If I was looking for a way to do DRY here then this is the code that I would factor out into a function. It is duplicated in every function here and in orthopolys.

oscarbenjamin commented on sympy/polys/appellseqs.py:
```diff
@@ -0,0 +1,295 @@
+    seq = [[1], [1, K(-1,2)]]
+    for i in range(2, n+1):
+        q = dup_integrate(dup_mul_ground(seq[-1], i, K), 1, K)
+        if i&1 ^ 1:
```
It took me a moment to parse `i&1 ^ 1`. Is this equivalent to `if i % 2 == 0` (i.e. just testing for an even number)? If so I think it's clearer to write it with `%` rather than bit-twiddling.

There doesn't seem to be any speed advantage for small integers:
```python
In [14]: n = 10

In [15]: %timeit n % 2 == 0
68.7 ns ± 0.242 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [16]: %timeit n & 1 ^ 1
120 ns ± 3.57 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
There is apparently a speed advantage for large integers but I can't believe that `n` here can be very large and any speed taken up in this operation will be dwarfed by the surrounding code.

Parcly-Taxel replied:
Done, done, done and done.

oscarbenjamin commented:
> polys
> * New `appellseqs` module generating Appell sequences, including the Bernoulli and Euler polynomials.

The release note should list the new public functions.

Parcly-Taxel committed:
Run orthopolys and appellseqs through a common interface

Including unifying the two Chebyshev generators into one function.
There are also two kinds of Hermite polynomials, and they too share the
same recurrence, but the second type He_n(x) (aka the probabilist,
reduced or small polynomials) will not be added here.

oscarbenjamin commented on sympy/polys/appellseqs.py:
```diff
@@ -0,0 +1,258 @@
+    """Low-level implementation of Bernoulli polynomials."""
+    if n < 1:
+        return [1]
+    p = [1, K(-1,2)]
```
When working with domain elements we should use `K.one` and `K.zero` instead of `1` and `0`.

oscarbenjamin replied:
See e.g. https://docs.sympy.org/latest/modules/polys/domainsintro.html#domain-elements-vs-sympy-expressions

oscarbenjamin commented on sympy/polys/numberfields/minpoly.py:
```diff
@@ -18,7 +18,7 @@
 
 from sympy.polys.domains import ZZ, QQ, FractionField
-from sympy.polys.orthopolys import dup_chebyshevt
+from sympy.polys.orthopolys import dup_chebyshev
```
I think the `dup_chebyshevt` and `dup_chebyshevu` functions should be preserved (although they could both call a new `dup_chebyshev` function).

Parcly-Taxel replied:
Both fixed.

oscarbenjamin commented:
I left a couple of comments but otherwise I think this looks good.

I notice that the line spacing in orthopolys is inconsistent with sometimes one and sometimes two blank lines between functions. Broadly throughout SymPy the convention is two blank lines between top level definitions but it seems that in polys the convention is more often one line so it's probably better to stick with that.

Parcly-Taxel committed:
Restore domain elements in dup_* functions

oscarbenjamin commented:
Looks good.
